Survival Games custom participant class

## Description
Convert Survival Games to use the new `TeamData` and `Participant` classes (make custom implementations) to store things like kills. 

## Definition of Done
- [x] Survival Games uses a custom `TeamData` and `Participant` implementation to store data about each player, rather than a series of parallel lists and maps. 

## Related Issues
- #572

```
[02:49:37 INFO]: 0:20 grace period
[02:49:57 INFO]: Grace period ended
[02:50:08 INFO]: Braekpo1nt was slain by rstln
[02:50:08 INFO]: Purple has been eliminated.
[02:50:08 INFO]: Purple got third place!
[02:52:01 INFO]: Kbelleflash was slain by rstln
[02:52:01 INFO]: Orange has been eliminated.
[02:52:01 INFO]: Orange got second place!
[02:52:01 INFO]: Team Yellow wins!
[02:52:11 WARN]: [MCTManager] Task #684 for MCTManager v1.3.0 generated an exception
java.lang.RuntimeException: Error running Timer "null"
        at MCTManager-1.3.0.jar/org.braekpo1nt.mctmanager.ui.timer.Timer.run(Timer.java:196) ~[MCTManager-1.3.0.jar:?]
        at org.bukkit.craftbukkit.scheduler.CraftTask.run(CraftTask.java:86) ~[paper-1.21.1.jar:1.21.1-109-355b1cb]
        at org.bukkit.craftbukkit.scheduler.CraftScheduler.mainThreadHeartbeat(CraftScheduler.java:475) ~[paper-1.21.1.jar:1.21.1-109-355b1cb]
        at net.minecraft.server.MinecraftServer.tickChildren(MinecraftServer.java:1724) ~[paper-1.21.1.jar:1.21.1-109-355b1cb]
        at net.minecraft.server.dedicated.DedicatedServer.tickChildren(DedicatedServer.java:473) ~[paper-1.21.1.jar:1.21.1-109-355b1cb]
        at net.minecraft.server.MinecraftServer.tickServer(MinecraftServer.java:1596) ~[paper-1.21.1.jar:1.21.1-109-355b1cb]
        at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1302) ~[paper-1.21.1.jar:1.21.1-109-355b1cb]
        at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:329) ~[paper-1.21.1.jar:1.21.1-109-355b1cb]
        at java.base/java.lang.Thread.run(Thread.java:1583) ~[?:?]
Caused by: java.lang.NullPointerException: Cannot invoke "org.braekpo1nt.mctmanager.ui.glow.GlowManager$PlayerData.getPlayer()" because "viewerPlayerData" is null
        at MCTManager-1.3.0.jar/org.braekpo1nt.mctmanager.ui.glow.GlowManager.removePlayer(GlowManager.java:327) ~[MCTManager-1.3.0.jar:?]
        at MCTManager-1.3.0.jar/org.braekpo1nt.mctmanager.ui.glow.GlowManager.removePlayer(GlowManager.java:304) ~[MCTManager-1.3.0.jar:?]
        at MCTManager-1.3.0.jar/org.braekpo1nt.mctmanager.games.game.survivalgames.SurvivalGamesGame.resetParticipant(SurvivalGamesGame.java:210) ~[MCTManager-1.3.0.jar:?]
        at MCTManager-1.3.0.jar/org.braekpo1nt.mctmanager.games.game.survivalgames.states.GameOverState.resetParticipant(GameOverState.java:56) ~[MCTManager-1.3.0.jar:?]
        at MCTManager-1.3.0.jar/org.braekpo1nt.mctmanager.games.game.survivalgames.SurvivalGamesGame.stop(SurvivalGamesGame.java:226) ~[MCTManager-1.3.0.jar:?]
        at MCTManager-1.3.0.jar/org.braekpo1nt.mctmanager.games.game.survivalgames.states.GameOverState.lambda$new$0(GameOverState.java:35) ~[MCTManager-1.3.0.jar:?]
        at MCTManager-1.3.0.jar/org.braekpo1nt.mctmanager.ui.timer.Timer.onComplete(Timer.java:206) ~[MCTManager-1.3.0.jar:?]
        at MCTManager-1.3.0.jar/org.braekpo1nt.mctmanager.ui.timer.Timer.run(Timer.java:162) ~[MCTManager-1.3.0.jar:?]
        ... 8 more
```

--- a/src/main/java/org/braekpo1nt/mctmanager/games/game/footrace/FootRaceGame.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/games/game/footrace/FootRaceGame.java
@@ -77,7 +77,7 @@ public class FootRaceGame implements Listener, MCTGame, Configurable, Headerable
     /**
      * Holds the data for participants when they have quit mid-round
      */
-    private Map<UUID, QuitData> quitDatas = new HashMap<>();
+    private Map<UUID, FootRaceQuitData> quitDatas = new HashMap<>();
     /**
      * what place every participant is in at any given moment in the race
      */

--- a/src/main/java/org/braekpo1nt/mctmanager/games/game/footrace/FootRaceParticipant.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/games/game/footrace/FootRaceParticipant.java
@@ -1,13 +1,11 @@
 package org.braekpo1nt.mctmanager.games.game.footrace;
 
-import lombok.EqualsAndHashCode;
 import lombok.Getter;
 import lombok.Setter;
 import lombok.ToString;
 import org.braekpo1nt.mctmanager.participant.Participant;
 import org.jetbrains.annotations.NotNull;
 
-@EqualsAndHashCode(callSuper = true)
 @ToString(callSuper = true)
 @Getter
 @Setter

--- a/src/main/java/org/braekpo1nt/mctmanager/games/game/footrace/FootRaceParticipant.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/games/game/footrace/FootRaceParticipant.java
@@ -44,7 +44,7 @@ public class FootRaceParticipant extends Participant {
         this.placement = 0;
     }
     
-    public FootRaceParticipant(Participant participant, QuitData quitData) {
+    public FootRaceParticipant(Participant participant, FootRaceQuitData quitData) {
         super(participant);
         this.lapCooldown = System.currentTimeMillis();
         this.lap = quitData.getLap();
@@ -53,8 +53,8 @@ public class FootRaceParticipant extends Participant {
         this.placement = quitData.getPlacement();
     }
     
-    public QuitData getQuitData() {
-        return new QuitData(
+    public FootRaceQuitData getQuitData() {
+        return new FootRaceQuitData(
                 lap,
                 currentCheckpoint,
                 finished,

--- a/src/main/java/org/braekpo1nt/mctmanager/games/game/footrace/FootRaceQuitData.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/games/game/footrace/FootRaceQuitData.java
@@ -6,7 +6,7 @@ import lombok.Data;
  * Holds data for the participant when they leave the game
  */
 @Data
-public class QuitData {
+public class FootRaceQuitData {
     /**
      * The participant's current lap
      */

--- a/src/main/java/org/braekpo1nt/mctmanager/games/game/footrace/states/ActiveState.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/games/game/footrace/states/ActiveState.java
@@ -6,7 +6,7 @@ import net.kyori.adventure.text.format.NamedTextColor;
 import org.braekpo1nt.mctmanager.games.GameManager;
 import org.braekpo1nt.mctmanager.games.game.footrace.FootRaceGame;
 import org.braekpo1nt.mctmanager.games.game.footrace.FootRaceParticipant;
-import org.braekpo1nt.mctmanager.games.game.footrace.QuitData;
+import org.braekpo1nt.mctmanager.games.game.footrace.FootRaceQuitData;
 import org.braekpo1nt.mctmanager.games.game.footrace.config.FootRaceConfig;
 import org.braekpo1nt.mctmanager.games.utils.GameManagerUtils;
 import org.braekpo1nt.mctmanager.participant.Participant;
@@ -80,7 +80,7 @@ public class ActiveState implements FootRaceState {
     
     @Override
     public void onParticipantJoin(Participant newParticipant) {
-        QuitData quitData = context.getQuitDatas().remove(newParticipant.getUniqueId());
+        FootRaceQuitData quitData = context.getQuitDatas().remove(newParticipant.getUniqueId());
         if (quitData != null) {
             FootRaceParticipant rejoinedParticipant = new FootRaceParticipant(newParticipant, quitData);
             rejoinParticipant(rejoinedParticipant);

--- a/src/main/java/org/braekpo1nt/mctmanager/games/game/parkourpathway/ParkourParticipant.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/games/game/parkourpathway/ParkourParticipant.java
@@ -1,13 +1,11 @@
 package org.braekpo1nt.mctmanager.games.game.parkourpathway;
 
-import lombok.EqualsAndHashCode;
 import lombok.Getter;
 import lombok.Setter;
 import lombok.ToString;
 import org.braekpo1nt.mctmanager.participant.Participant;
 import org.jetbrains.annotations.NotNull;
 
-@EqualsAndHashCode(callSuper = true)
 @ToString(callSuper = true)
 @Getter
 @Setter

--- a/src/main/java/org/braekpo1nt/mctmanager/games/game/spleef/SpleefParticipant.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/games/game/spleef/SpleefParticipant.java
@@ -4,7 +4,6 @@ import lombok.*;
 import org.braekpo1nt.mctmanager.participant.Participant;
 import org.jetbrains.annotations.NotNull;
 
-@EqualsAndHashCode(callSuper = true)
 @ToString(callSuper = true)
 @Getter
 @Setter

--- a/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/SurvivalGamesGame.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/SurvivalGamesGame.java
@@ -169,7 +169,7 @@ public class SurvivalGamesGame implements MCTGame, Configurable, Listener, Heade
      * @param participant the participant to show their teammates to
      */
     public void initializeGlowing(Participant participant) {
-        for (Participant other : participants.values()) {
+        for (SurvivalGamesParticipant other : participants.values()) {
             if (!other.equals(participant)) {
                 if (participant.getTeamId().equals(other.getTeamId())) {
                     glowManager.showGlowing(participant, other);

--- a/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/SurvivalGamesGame.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/SurvivalGamesGame.java
@@ -191,7 +191,7 @@ public class SurvivalGamesGame implements MCTGame, Configurable, Listener, Heade
         sidebar.addPlayer(participant);
         topbar.showPlayer(participant);
         topbar.linkToTeam(participant.getUniqueId(), teamId);
-        glowManager.addPlayer(participant);//
+        glowManager.addPlayer(participant);
         updateAliveCount(team);
         initializeKillCount(participant);
         participant.setGameMode(GameMode.ADVENTURE);

--- a/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/SurvivalGamesGame.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/SurvivalGamesGame.java
@@ -254,8 +254,12 @@ public class SurvivalGamesGame implements MCTGame, Configurable, Listener, Heade
     
     @Override
     public void onParticipantQuit(Participant participant, Team team) {
+        SurvivalGamesParticipant sgParticipant = participants.get(participant.getUniqueId());
+        if (sgParticipant == null) {
+            return;
+        }
         if (state != null) {
-            state.onParticipantQuit(participant);
+            state.onParticipantQuit(sgParticipant);
         }
     }
     

--- a/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/SurvivalGamesGame.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/SurvivalGamesGame.java
@@ -4,7 +4,6 @@ import lombok.Data;
 import net.kyori.adventure.audience.Audience;
 import net.kyori.adventure.text.Component;
 import net.kyori.adventure.text.format.NamedTextColor;
-import net.kyori.adventure.text.format.TextColor;
 import org.braekpo1nt.mctmanager.Main;
 import org.braekpo1nt.mctmanager.config.exceptions.ConfigIOException;
 import org.braekpo1nt.mctmanager.config.exceptions.ConfigInvalidException;
@@ -19,7 +18,6 @@ import org.braekpo1nt.mctmanager.games.game.survivalgames.states.SurvivalGamesSt
 import org.braekpo1nt.mctmanager.games.utils.ParticipantInitializer;
 import org.braekpo1nt.mctmanager.participant.Participant;
 import org.braekpo1nt.mctmanager.participant.Team;
-import org.braekpo1nt.mctmanager.participant.TeamData;
 import org.braekpo1nt.mctmanager.ui.glow.GlowManager;
 import org.braekpo1nt.mctmanager.ui.sidebar.Headerable;
 import org.braekpo1nt.mctmanager.ui.sidebar.KeyLine;
@@ -76,18 +74,11 @@ public class SurvivalGamesGame implements MCTGame, Configurable, Listener, Heade
     private Sidebar sidebar;
     private Sidebar adminSidebar;
     private SurvivalGamesConfig config;
-    public Map<String, TeamData<Participant>> teams = new HashMap<>();
-    public Map<UUID, Participant> participants = new HashMap<>();
+    public Map<String, SurvivalGamesTeam> teams = new HashMap<>();
+    public Map<UUID, SurvivalGamesParticipant> participants = new HashMap<>();
+    public Map<UUID, SurvivalGamesQuitData> quitDatas = new HashMap<>();
     private List<Player> admins = new ArrayList<>();
     private WorldBorder worldBorder;
-    private List<UUID> livingPlayers = new ArrayList<>();
-    /**
-     * a map of all teamIds to how many members are alive on that team
-     */
-    private Map<String, Integer> livingMembers = new HashMap<>();
-    private List<UUID> deadPlayers = new ArrayList<>();
-    private Map<UUID, Integer> killCounts = new HashMap<>();
-    private Map<UUID, Integer> deathCounts = new HashMap<>();
     private Component title = baseTitle;
     
     public SurvivalGamesGame(Main plugin, GameManager gameManager) {
@@ -129,19 +120,14 @@ public class SurvivalGamesGame implements MCTGame, Configurable, Listener, Heade
     public void start(Collection<Team> newTeams, Collection<Participant> newParticipants, List<Player> newAdmins) {
         this.teams = new HashMap<>(newTeams.size());
         for (Team team : newTeams) {
-            teams.put(team.getTeamId(), new TeamData<>(team));
+            teams.put(team.getTeamId(), new SurvivalGamesTeam(team));
+            topbar.addTeam(team.getTeamId(), team.getColor());
         }
         this.participants = new HashMap<>(newParticipants.size());
-        livingPlayers = new ArrayList<>(newParticipants.size());
-        deadPlayers = new ArrayList<>();
-        livingMembers = new HashMap<>(Participant.getTeamIds(newParticipants).size());
-        killCounts = new HashMap<>(newParticipants.size());
-        deathCounts = new HashMap<>(newParticipants.size());
+        this.quitDatas = new HashMap<>();
         worldBorder = config.getWorld().getWorldBorder();
         sidebar = gameManager.createSidebar();
         adminSidebar = gameManager.createSidebar();
-        Set<String> teams = Participant.getTeamIds(newParticipants);
-        setUpTopbarTeams(teams);
         plugin.getServer().getPluginManager().registerEvents(this, plugin);
         glowManager.registerListeners();
         gameManager.getTimerManager().register(timerManager);
@@ -196,18 +182,17 @@ public class SurvivalGamesGame implements MCTGame, Configurable, Listener, Heade
         }
     }
     
-    public void initializeParticipant(Participant participant) {
+    public void initializeParticipant(Participant newParticipant) {
+        SurvivalGamesParticipant participant = new SurvivalGamesParticipant(newParticipant);
         participants.put(participant.getUniqueId(), participant);
-        livingPlayers.add(participant.getUniqueId());
+        SurvivalGamesTeam team = teams.get(participant.getTeamId());
+        team.addParticipant(participant);
         String teamId = participant.getTeamId();
-        livingMembers.putIfAbsent(teamId, 0);
-        int oldAliveCount = livingMembers.get(teamId);
-        livingMembers.put(teamId, oldAliveCount + 1);
         sidebar.addPlayer(participant);
         topbar.showPlayer(participant);
         topbar.linkToTeam(participant.getUniqueId(), teamId);
         glowManager.addPlayer(participant);
-        updateAliveCount(teamId);
+        updateAliveCount(team);
         initializeKillCount(participant);
         participant.setGameMode(GameMode.ADVENTURE);
         ParticipantInitializer.clearInventory(participant);
@@ -223,13 +208,6 @@ public class SurvivalGamesGame implements MCTGame, Configurable, Listener, Heade
         sidebar.removePlayer(participant.getUniqueId());
         topbar.hidePlayer(participant.getUniqueId());
         glowManager.removePlayer(participant);
-    }
-    
-    private void setUpTopbarTeams(Set<String> newTeamIds) {
-        for (String teamId : newTeamIds) {
-            TextColor color = gameManager.getTeamColor(teamId);
-            topbar.addTeam(teamId, color);
-        }
     }
     
     @Override
@@ -254,11 +232,7 @@ public class SurvivalGamesGame implements MCTGame, Configurable, Listener, Heade
         glowManager.clear();
         teams.clear();
         participants.clear();
-        livingPlayers.clear();
-        deadPlayers.clear();
-        livingMembers.clear();
-        killCounts.clear();
-        deathCounts.clear();
+        quitDatas.clear();
         gameManager.gameIsOver();
         state = null;
         Main.logger().info("Stopped Survival Games");
@@ -268,7 +242,7 @@ public class SurvivalGamesGame implements MCTGame, Configurable, Listener, Heade
         if (teams.containsKey(team.getTeamId())) {
             return;
         }
-        teams.put(team.getTeamId(), new TeamData<>(team));
+        teams.put(team.getTeamId(), new SurvivalGamesTeam(team));
     }
     
     @Override
@@ -466,41 +440,20 @@ public class SurvivalGamesGame implements MCTGame, Configurable, Listener, Heade
     
     /**
      * updates the number of living and dead players in the topbar
-     * @param teamId the team to update
+     * @param team the team to update
      */
-    public void updateAliveCount(@NotNull String teamId) {
-        int alive = livingMembers.get(teamId);
-        int dead = getDeadMembers(teamId);
-        topbar.setMembers(teamId, alive, dead);
+    public void updateAliveCount(@NotNull SurvivalGamesTeam team) {
+        topbar.setMembers(team.getTeamId(), team.getAlive(), team.getDead());
     }
     
-    /**
-     * @param teamId the teamId of the team to count the dead members of
-     * @return the number of dead players on the given team in this game
-     */
-    private int getDeadMembers(String teamId) {
-        int count = 0;
-        for (Participant participant : participants.values()) {
-            if (deadPlayers.contains(participant.getUniqueId())) {
-                if (teamId.equals(participant.getTeamId())) {
-                    count++;
-                }
-            }
-        }
-        return count;
-    }
-    
-    public void initializeKillCount(Participant participant) {
-        killCounts.putIfAbsent(participant.getUniqueId(), 0);
-        deathCounts.putIfAbsent(participant.getUniqueId(), 0);
-        int kills = killCounts.get(participant.getUniqueId());
+    public void initializeKillCount(SurvivalGamesParticipant participant) {
         int deaths;
         if (!config.showDeathCount()) {
             deaths = -1;
         } else {
-            deaths = deathCounts.get(participant.getUniqueId());
+            deaths = participant.getDeaths();
         }
-        topbar.setKillsAndDeaths(participant.getUniqueId(), kills, deaths);
+        topbar.setKillsAndDeaths(participant.getUniqueId(), participant.getKills(), deaths);
     }
     
     private void cancelAllTasks() {
@@ -785,9 +738,11 @@ public class SurvivalGamesGame implements MCTGame, Configurable, Listener, Heade
      * @return the number of {@link HumanEntity}s in the viewers list who are living participants
      */
     private long countLivingParticipantViewers(List<HumanEntity> viewers) {
-        return viewers.stream().filter(viewer -> 
-                participants.containsKey(viewer.getUniqueId()) 
-                && livingPlayers.contains(viewer.getUniqueId())).count();
+        return viewers.stream()
+                .map(viewer -> participants.get(viewer.getUniqueId()))
+                .filter(Objects::nonNull)
+                .filter(SurvivalGamesParticipant::isAlive)
+                .count();
     }
     
     // State-specific callers

--- a/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/SurvivalGamesGame.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/SurvivalGamesGame.java
@@ -191,7 +191,7 @@ public class SurvivalGamesGame implements MCTGame, Configurable, Listener, Heade
         sidebar.addPlayer(participant);
         topbar.showPlayer(participant);
         topbar.linkToTeam(participant.getUniqueId(), teamId);
-        glowManager.addPlayer(participant);
+        glowManager.addPlayer(participant);//
         updateAliveCount(team);
         initializeKillCount(participant);
         participant.setGameMode(GameMode.ADVENTURE);

--- a/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/SurvivalGamesParticipant.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/SurvivalGamesParticipant.java
@@ -1,0 +1,23 @@
+package org.braekpo1nt.mctmanager.games.game.survivalgames;
+
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.ToString;
+import org.braekpo1nt.mctmanager.participant.Participant;
+import org.jetbrains.annotations.NotNull;
+
+@EqualsAndHashCode(callSuper = true)
+@ToString(callSuper = true)
+@Getter
+@Setter
+public class SurvivalGamesParticipant extends Participant {
+    
+    private boolean alive;
+    private int kills;
+    private int deaths;
+    
+    public SurvivalGamesParticipant(@NotNull Participant participant) {
+        super(participant);
+    }
+}

--- a/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/SurvivalGamesParticipant.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/SurvivalGamesParticipant.java
@@ -1,13 +1,11 @@
 package org.braekpo1nt.mctmanager.games.game.survivalgames;
 
-import lombok.EqualsAndHashCode;
 import lombok.Getter;
 import lombok.Setter;
 import lombok.ToString;
 import org.braekpo1nt.mctmanager.participant.Participant;
 import org.jetbrains.annotations.NotNull;
 
-@EqualsAndHashCode(callSuper = true)
 @ToString(callSuper = true)
 @Getter
 @Setter

--- a/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/SurvivalGamesParticipant.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/SurvivalGamesParticipant.java
@@ -19,5 +19,23 @@ public class SurvivalGamesParticipant extends Participant {
     
     public SurvivalGamesParticipant(@NotNull Participant participant) {
         super(participant);
+        this.alive = true;
+        this.kills = 0;
+        this.deaths = 0;
+    }
+    
+    public SurvivalGamesParticipant(@NotNull Participant participant, @NotNull SurvivalGamesQuitData quitData) {
+        super(participant);
+        this.alive = quitData.isAlive();
+        this.kills = quitData.getKills();
+        this.deaths = quitData.getDeaths();
+    }
+    
+    public SurvivalGamesQuitData getQuitData() {
+        return new SurvivalGamesQuitData(
+                alive,
+                kills,
+                deaths
+        );
     }
 }

--- a/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/SurvivalGamesQuitData.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/SurvivalGamesQuitData.java
@@ -1,0 +1,10 @@
+package org.braekpo1nt.mctmanager.games.game.survivalgames;
+
+import lombok.Data;
+
+@Data
+public class SurvivalGamesQuitData {
+    private final boolean alive;
+    private final int kills;
+    private final int deaths;
+}

--- a/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/SurvivalGamesTeam.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/SurvivalGamesTeam.java
@@ -23,4 +23,12 @@ public class SurvivalGamesTeam extends TeamData<SurvivalGamesParticipant> {
     public boolean isAlive() {
         return getParticipants().stream().anyMatch(SurvivalGamesParticipant::isAlive);
     }
+    
+    public int getAlive() {
+        return (int) getParticipants().stream().filter(SurvivalGamesParticipant::isAlive).count();
+    }
+    
+    public int getDead() {
+        return (int) getParticipants().stream().filter(p -> !p.isAlive()).count();
+    }
 }

--- a/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/SurvivalGamesTeam.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/SurvivalGamesTeam.java
@@ -1,0 +1,26 @@
+package org.braekpo1nt.mctmanager.games.game.survivalgames;
+
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.ToString;
+import org.braekpo1nt.mctmanager.participant.Team;
+import org.braekpo1nt.mctmanager.participant.TeamData;
+
+@EqualsAndHashCode(callSuper = true)
+@ToString(callSuper = true)
+@Getter
+@Setter
+public class SurvivalGamesTeam extends TeamData<SurvivalGamesParticipant> {
+    
+    public SurvivalGamesTeam(Team team) {
+        super(team);
+    }
+    
+    /**
+     * @return True if at least one participant on this team is alive, false otherwise
+     */
+    public boolean isAlive() {
+        return getParticipants().stream().anyMatch(SurvivalGamesParticipant::isAlive);
+    }
+}

--- a/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/states/ActiveState.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/states/ActiveState.java
@@ -188,7 +188,7 @@ public class ActiveState implements SurvivalGamesState {
     @Override
     public void onParticipantJoin(Participant participant, Team team) {
         context.onTeamJoin(team);
-        SurvivalGamesQuitData quitData = context.getQuitDatas().get(participant.getUniqueId());
+        SurvivalGamesQuitData quitData = context.getQuitDatas().remove(participant.getUniqueId());
         if (quitData != null) {
             SurvivalGamesParticipant sgParticipant = new SurvivalGamesParticipant(
                     participant,
@@ -232,6 +232,7 @@ public class ActiveState implements SurvivalGamesState {
         }
         resetParticipant(participant);
         context.getParticipants().remove(participant.getUniqueId());
+        context.getQuitDatas().put(participant.getUniqueId(), participant.getQuitData());
     }
     
     @Override
@@ -337,7 +338,6 @@ public class ActiveState implements SurvivalGamesState {
         context.messageAllParticipants(Component.empty()
                 .append(deadTeam.getFormattedDisplayName())
                 .append(Component.text(" has been eliminated.")));
-        // TODO: does getLivingTeams need to return teams, or just ids?
         List<SurvivalGamesTeam> livingTeams = getLivingTeams();
         gameManager.awardPointsToTeams(Team.getTeamIds(livingTeams), config.getSurviveTeamScore());
         switch (livingTeams.size()) {

--- a/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/states/ActiveState.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/states/ActiveState.java
@@ -3,18 +3,17 @@ package org.braekpo1nt.mctmanager.games.game.survivalgames.states;
 import net.kyori.adventure.audience.Audience;
 import net.kyori.adventure.text.Component;
 import net.kyori.adventure.text.format.NamedTextColor;
-import net.kyori.adventure.text.format.TextColor;
 import org.braekpo1nt.mctmanager.Main;
 import org.braekpo1nt.mctmanager.games.GameManager;
-import org.braekpo1nt.mctmanager.games.game.survivalgames.SurvivalGamesGame;
 import org.braekpo1nt.mctmanager.games.game.survivalgames.SurvivalGamesParticipant;
+import org.braekpo1nt.mctmanager.games.game.survivalgames.SurvivalGamesGame;
+import org.braekpo1nt.mctmanager.games.game.survivalgames.SurvivalGamesQuitData;
 import org.braekpo1nt.mctmanager.games.game.survivalgames.SurvivalGamesTeam;
 import org.braekpo1nt.mctmanager.games.game.survivalgames.config.SurvivalGamesConfig;
 import org.braekpo1nt.mctmanager.games.utils.GameManagerUtils;
 import org.braekpo1nt.mctmanager.games.utils.ParticipantInitializer;
 import org.braekpo1nt.mctmanager.participant.Participant;
 import org.braekpo1nt.mctmanager.participant.Team;
-import org.braekpo1nt.mctmanager.participant.TeamData;
 import org.braekpo1nt.mctmanager.ui.TimeStringUtils;
 import org.braekpo1nt.mctmanager.ui.UIUtils;
 import org.braekpo1nt.mctmanager.ui.sidebar.Sidebar;
@@ -189,14 +188,15 @@ public class ActiveState implements SurvivalGamesState {
     @Override
     public void onParticipantJoin(Participant participant, Team team) {
         context.onTeamJoin(team);
-        if (participantShouldRejoin(participant)) {
-            rejoinParticipant(participant);
+        SurvivalGamesQuitData quitData = context.getQuitDatas().get(participant.getUniqueId());
+        if (quitData != null) {
+            SurvivalGamesParticipant sgParticipant = new SurvivalGamesParticipant(
+                    participant,
+                    quitData
+            );
+            rejoinParticipant(sgParticipant);
         } else {
-            context.getDeadPlayers().remove(participant.getUniqueId());
-            String teamId = participant.getTeamId();
-            if (!context.getLivingMembers().containsKey(teamId)) {
-                context.getTopbar().addTeam(teamId, team.getColor());
-            }
+            context.getTopbar().addTeam(participant.getTeamId(), team.getColor());
             context.initializeParticipant(participant);
             participant.teleport(config.getPlatformSpawns().getFirst());
             participant.setRespawnLocation(config.getPlatformSpawns().getFirst(), true);
@@ -205,12 +205,7 @@ public class ActiveState implements SurvivalGamesState {
         context.initializeGlowing(participant);
     }
     
-    private boolean participantShouldRejoin(Participant participant) {
-        return context.getLivingMembers().containsKey(participant.getTeamId()) && 
-                context.getDeadPlayers().contains(participant.getUniqueId());
-    }
-    
-    private void rejoinParticipant(Participant participant) {
+    private void rejoinParticipant(SurvivalGamesParticipant participant) {
         context.getParticipants().put(participant.getUniqueId(), participant);
         context.getTeams().get(participant.getTeamId()).addParticipant(participant);
         participant.setGameMode(GameMode.SPECTATOR);
@@ -222,8 +217,8 @@ public class ActiveState implements SurvivalGamesState {
     }
     
     @Override
-    public void onParticipantQuit(Participant participant) {
-        if (context.getLivingPlayers().contains(participant.getUniqueId())) {
+    public void onParticipantQuit(SurvivalGamesParticipant participant) {
+        if (participant.isAlive()) {
             List<ItemStack> drops = Arrays.stream(participant.getInventory().getContents())
                     .filter(Objects::nonNull)
                     .toList();
@@ -263,7 +258,7 @@ public class ActiveState implements SurvivalGamesState {
     
     @Override
     public void onParticipantDeath(PlayerDeathEvent event) {
-        Participant killed = context.getParticipants().get(event.getPlayer().getUniqueId());
+        SurvivalGamesParticipant killed = context.getParticipants().get(event.getPlayer().getUniqueId());
         if (killed == null) {
             return;
         }
@@ -279,7 +274,7 @@ public class ActiveState implements SurvivalGamesState {
             plugin.getServer().sendMessage(deathMessage);
         }
         if (killed.getKiller() != null) {
-            Participant killer = context.getParticipants().get(killed.getKiller().getUniqueId());
+            SurvivalGamesParticipant killer = context.getParticipants().get(killed.getKiller().getUniqueId());
             if (killer != null) {
                 onParticipantGetKill(killer, killed);
             }
@@ -293,64 +288,57 @@ public class ActiveState implements SurvivalGamesState {
         }
     }
     
-    private void onParticipantGetKill(@NotNull Participant killer, @NotNull Participant killed) {
+    private void onParticipantGetKill(@NotNull SurvivalGamesParticipant killer, @NotNull SurvivalGamesParticipant killed) {
         if (!context.getParticipants().containsKey(killer.getUniqueId())) {
             return;
         }
-        addKill(killer.getUniqueId());
+        addKill(killer);
         UIUtils.showKillTitle(killer, killed);
         gameManager.awardPointsToParticipant(killer, config.getKillScore());
     }
     
     /**
-     * @param playerUUID the player to add a kill to
+     * @param participant the participant to add a kill to
      */
-    private void addKill(@NotNull UUID playerUUID) {
-        int oldKillCount = context.getKillCounts().get(playerUUID);
+    private void addKill(@NotNull SurvivalGamesParticipant participant) {
+        int oldKillCount = participant.getKills();
         int newKillCount = oldKillCount + 1;
-        context.getKillCounts().put(playerUUID, newKillCount);
-        topbar.setKills(playerUUID, newKillCount);
+        participant.setKills(newKillCount);
+        topbar.setKills(participant.getUniqueId(), newKillCount);
     }
     
     /**
-     * @param playerUUID the player to add a death to
+     * @param participant the participant to add a death to
      */
-    private void addDeath(@NotNull UUID playerUUID) {
-        int oldDeathCount = context.getDeathCounts().get(playerUUID);
+    private void addDeath(@NotNull SurvivalGamesParticipant participant) {
+        int oldDeathCount = participant.getDeaths();
         int newDeathCount = oldDeathCount + 1;
-        context.getDeathCounts().put(playerUUID, newDeathCount);
-        topbar.setDeaths(playerUUID, newDeathCount);
+        participant.setDeaths(newDeathCount);
+        topbar.setDeaths(participant.getUniqueId(), newDeathCount);
     }
     
-    private void onParticipantDeath(Participant killed) {
-        UUID killedUUID = killed.getUniqueId();
+    private void onParticipantDeath(SurvivalGamesParticipant killed) {
         killed.getInventory().clear();
-        switchPlayerFromLivingToDead(killedUUID);
+        killed.setAlive(false);
         String teamId = killed.getTeamId();
-        int oldLivingMembers = context.getLivingMembers().get(teamId);
-        context.getLivingMembers().put(teamId, oldLivingMembers - 1);
-        addDeath(killedUUID);
-        context.updateAliveCount(teamId);
-        if (context.getLivingMembers().get(teamId) <= 0) {
+        SurvivalGamesTeam team = context.getTeams().get(teamId);
+        addDeath(killed);
+        context.updateAliveCount(team);
+        if (!team.isAlive()) {
             onTeamDeath(context.getTeams().get(teamId));
         }
-    }
-    
-    private void switchPlayerFromLivingToDead(UUID playerUniqueId) {
-        context.getLivingPlayers().remove(playerUniqueId);
-        context.getDeadPlayers().add(playerUniqueId);
     }
     
     /**
      * Call when all of a team's members are dead. 
      * @param deadTeam the team who just died
      */
-    private void onTeamDeath(Team deadTeam) {
+    private void onTeamDeath(SurvivalGamesTeam deadTeam) {
         context.messageAllParticipants(Component.empty()
                 .append(deadTeam.getFormattedDisplayName())
                 .append(Component.text(" has been eliminated.")));
         // TODO: does getLivingTeams need to return teams, or just ids?
-        List<TeamData<Participant>> livingTeams = getLivingTeams();
+        List<SurvivalGamesTeam> livingTeams = getLivingTeams();
         gameManager.awardPointsToTeams(Team.getTeamIds(livingTeams), config.getSurviveTeamScore());
         switch (livingTeams.size()) {
             case 2 -> {
@@ -377,14 +365,12 @@ public class ActiveState implements SurvivalGamesState {
      * @return a list of the teamIds of the teams which are still alive (have at least 1 living member)
      */
     private @NotNull List<SurvivalGamesTeam> getLivingTeams() {
-        // TODO: Teams remove this in favor of teams storing their living members
-        return context.getLivingMembers().entrySet().stream()
-                .filter(entry -> entry.getValue() > 0)
-                .map(Map.Entry::getKey)
-                .map(context.getTeams()::get).toList();
+        return context.getTeams().values().stream()
+                .filter(SurvivalGamesTeam::isAlive)
+                .toList();
     }
     
-    private void onTeamWin(Team winningTeam) {
+    private void onTeamWin(SurvivalGamesTeam winningTeam) {
         plugin.getServer().sendMessage(Component.text("Team ")
                 .append(winningTeam.getFormattedDisplayName())
                 .append(Component.text(" wins!")));

--- a/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/states/ActiveState.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/states/ActiveState.java
@@ -7,6 +7,7 @@ import net.kyori.adventure.text.format.TextColor;
 import org.braekpo1nt.mctmanager.Main;
 import org.braekpo1nt.mctmanager.games.GameManager;
 import org.braekpo1nt.mctmanager.games.game.survivalgames.SurvivalGamesGame;
+import org.braekpo1nt.mctmanager.games.game.survivalgames.SurvivalGamesTeam;
 import org.braekpo1nt.mctmanager.games.game.survivalgames.config.SurvivalGamesConfig;
 import org.braekpo1nt.mctmanager.games.utils.GameManagerUtils;
 import org.braekpo1nt.mctmanager.games.utils.ParticipantInitializer;
@@ -394,7 +395,7 @@ public class ActiveState implements SurvivalGamesState {
     /**
      * @return a list of the teamIds of the teams which are still alive (have at least 1 living member)
      */
-    private @NotNull List<TeamData<Participant>> getLivingTeams() {
+    private @NotNull List<SurvivalGamesTeam> getLivingTeams() {
         // TODO: Teams remove this in favor of teams storing their living members
         return context.getLivingMembers().entrySet().stream()
                 .filter(entry -> entry.getValue() > 0)

--- a/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/states/ActiveState.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/states/ActiveState.java
@@ -7,6 +7,7 @@ import net.kyori.adventure.text.format.TextColor;
 import org.braekpo1nt.mctmanager.Main;
 import org.braekpo1nt.mctmanager.games.GameManager;
 import org.braekpo1nt.mctmanager.games.game.survivalgames.SurvivalGamesGame;
+import org.braekpo1nt.mctmanager.games.game.survivalgames.SurvivalGamesParticipant;
 import org.braekpo1nt.mctmanager.games.game.survivalgames.SurvivalGamesTeam;
 import org.braekpo1nt.mctmanager.games.game.survivalgames.config.SurvivalGamesConfig;
 import org.braekpo1nt.mctmanager.games.utils.GameManagerUtils;
@@ -196,7 +197,7 @@ public class ActiveState implements SurvivalGamesState {
             if (!context.getLivingMembers().containsKey(teamId)) {
                 context.getTopbar().addTeam(teamId, team.getColor());
             }
-            initializeParticipant(participant);
+            context.initializeParticipant(participant);
             participant.teleport(config.getPlatformSpawns().getFirst());
             participant.setRespawnLocation(config.getPlatformSpawns().getFirst(), true);
         }
@@ -236,26 +237,6 @@ public class ActiveState implements SurvivalGamesState {
         }
         resetParticipant(participant);
         context.getParticipants().remove(participant.getUniqueId());
-    }
-    
-    @Override
-    public void initializeParticipant(Participant participant) {
-        context.getParticipants().put(participant.getUniqueId(), participant);
-        context.getTeams().get(participant.getTeamId()).addParticipant(participant);
-        context.getLivingPlayers().add(participant.getUniqueId());
-        String teamId = participant.getTeamId();
-        context.getLivingMembers().putIfAbsent(teamId, 0);
-        int oldAliveCount = context.getLivingMembers().get(teamId);
-        context.getLivingMembers().put(teamId, oldAliveCount + 1);
-        sidebar.addPlayer(participant);
-        topbar.showPlayer(participant);
-        topbar.linkToTeam(participant.getUniqueId(), teamId);
-        context.updateAliveCount(teamId);
-        context.initializeKillCount(participant);
-        participant.setGameMode(GameMode.ADVENTURE);
-        ParticipantInitializer.clearInventory(participant);
-        ParticipantInitializer.resetHealthAndHunger(participant);
-        ParticipantInitializer.clearStatusEffects(participant);
     }
     
     @Override

--- a/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/states/DescriptionState.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/states/DescriptionState.java
@@ -3,6 +3,7 @@ package org.braekpo1nt.mctmanager.games.game.survivalgames.states;
 import net.kyori.adventure.text.Component;
 import org.braekpo1nt.mctmanager.Main;
 import org.braekpo1nt.mctmanager.games.game.survivalgames.SurvivalGamesGame;
+import org.braekpo1nt.mctmanager.games.game.survivalgames.SurvivalGamesParticipant;
 import org.braekpo1nt.mctmanager.games.utils.GameManagerUtils;
 import org.braekpo1nt.mctmanager.games.utils.ParticipantInitializer;
 import org.braekpo1nt.mctmanager.participant.Participant;
@@ -50,7 +51,7 @@ public class DescriptionState implements SurvivalGamesState {
     public void onParticipantJoin(Participant participant, Team team) {
         onTeamJoin(team);
         context.getDeadPlayers().remove(participant.getUniqueId());
-        initializeParticipant(participant);
+        context.initializeParticipant(participant);
         context.createPlatformsAndTeleportTeams();
         context.getSidebar().updateLine(participant.getUniqueId(), "title", context.getTitle());
         context.initializeGlowing(participant);
@@ -71,27 +72,6 @@ public class DescriptionState implements SurvivalGamesState {
         context.getDeathCounts().remove(participantUUID);
         context.getTopbar().unlinkFromTeam(participantUUID);
         resetParticipant(participant);
-    }
-    
-    @Override
-    public void initializeParticipant(Participant participant) {
-        context.getParticipants().put(participant.getUniqueId(), participant);
-        context.getTeams().get(participant.getTeamId()).addParticipant(participant);
-        context.getLivingPlayers().add(participant.getUniqueId());
-        String teamId = participant.getTeamId();
-        context.getLivingMembers().putIfAbsent(teamId, 0);
-        int oldAliveCount = context.getLivingMembers().get(teamId);
-        context.getLivingMembers().put(teamId, oldAliveCount + 1);
-        context.getSidebar().addPlayer(participant);
-        context.getTopbar().showPlayer(participant);
-        context.getTopbar().linkToTeam(participant.getUniqueId(), teamId);
-        context.updateAliveCount(teamId);
-        context.initializeKillCount(participant);
-        participant.setGameMode(GameMode.ADVENTURE);
-        ParticipantInitializer.clearInventory(participant);
-        ParticipantInitializer.resetHealthAndHunger(participant);
-        ParticipantInitializer.clearStatusEffects(participant);
-        context.getGlowManager().addPlayer(participant);
     }
     
     @Override

--- a/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/states/DescriptionState.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/states/DescriptionState.java
@@ -11,7 +11,6 @@ import org.braekpo1nt.mctmanager.participant.Team;
 import org.braekpo1nt.mctmanager.participant.TeamData;
 import org.braekpo1nt.mctmanager.ui.timer.Timer;
 import org.braekpo1nt.mctmanager.utils.LogType;
-import org.bukkit.GameMode;
 import org.bukkit.event.entity.EntityDamageEvent;
 import org.bukkit.event.entity.PlayerDeathEvent;
 import org.jetbrains.annotations.NotNull;
@@ -58,7 +57,7 @@ public class DescriptionState implements SurvivalGamesState {
     }
     
     @Override
-    public void onParticipantQuit(Participant participant) {
+    public void onParticipantQuit(SurvivalGamesParticipant participant) {
         context.getParticipants().remove(participant.getUniqueId());
         UUID participantUUID = participant.getUniqueId();
         String teamId = participant.getTeamId();

--- a/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/states/DescriptionState.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/states/DescriptionState.java
@@ -4,18 +4,16 @@ import net.kyori.adventure.text.Component;
 import org.braekpo1nt.mctmanager.Main;
 import org.braekpo1nt.mctmanager.games.game.survivalgames.SurvivalGamesGame;
 import org.braekpo1nt.mctmanager.games.game.survivalgames.SurvivalGamesParticipant;
+import org.braekpo1nt.mctmanager.games.game.survivalgames.SurvivalGamesTeam;
 import org.braekpo1nt.mctmanager.games.utils.GameManagerUtils;
 import org.braekpo1nt.mctmanager.games.utils.ParticipantInitializer;
 import org.braekpo1nt.mctmanager.participant.Participant;
 import org.braekpo1nt.mctmanager.participant.Team;
-import org.braekpo1nt.mctmanager.participant.TeamData;
 import org.braekpo1nt.mctmanager.ui.timer.Timer;
 import org.braekpo1nt.mctmanager.utils.LogType;
 import org.bukkit.event.entity.EntityDamageEvent;
 import org.bukkit.event.entity.PlayerDeathEvent;
 import org.jetbrains.annotations.NotNull;
-
-import java.util.UUID;
 
 public class DescriptionState implements SurvivalGamesState {
     
@@ -41,7 +39,7 @@ public class DescriptionState implements SurvivalGamesState {
         if (context.getTeams().containsKey(team.getTeamId())) {
             return;
         }
-        context.getTeams().put(team.getTeamId(), new TeamData<>(team));
+        context.getTeams().put(team.getTeamId(), new SurvivalGamesTeam(team));
         context.createPlatformsAndTeleportTeams();
         context.getTopbar().addTeam(team.getTeamId(), team.getColor());
     }
@@ -49,7 +47,6 @@ public class DescriptionState implements SurvivalGamesState {
     @Override
     public void onParticipantJoin(Participant participant, Team team) {
         onTeamJoin(team);
-        context.getDeadPlayers().remove(participant.getUniqueId());
         context.initializeParticipant(participant);
         context.createPlatformsAndTeleportTeams();
         context.getSidebar().updateLine(participant.getUniqueId(), "title", context.getTitle());
@@ -59,17 +56,9 @@ public class DescriptionState implements SurvivalGamesState {
     @Override
     public void onParticipantQuit(SurvivalGamesParticipant participant) {
         context.getParticipants().remove(participant.getUniqueId());
-        UUID participantUUID = participant.getUniqueId();
-        String teamId = participant.getTeamId();
-        Integer oldLivingMembers = context.getLivingMembers().get(teamId);
-        if (oldLivingMembers != null) {
-            context.getLivingMembers().put(teamId, Math.max(0, oldLivingMembers - 1));
-            context.updateAliveCount(teamId);
-        }
-        context.getLivingPlayers().remove(participantUUID);
-        context.getKillCounts().remove(participantUUID);
-        context.getDeathCounts().remove(participantUUID);
-        context.getTopbar().unlinkFromTeam(participantUUID);
+        SurvivalGamesTeam team = context.getTeams().get(participant.getTeamId());
+        context.updateAliveCount(team);
+        context.getTopbar().unlinkFromTeam(participant.getUniqueId());
         resetParticipant(participant);
     }
     

--- a/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/states/GameOverState.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/states/GameOverState.java
@@ -39,6 +39,7 @@ public class GameOverState implements SurvivalGamesState {
     
     @Override
     public void onParticipantJoin(Participant participant, Team team) {
+        context.getGlowManager().addPlayer(participant);
         context.initializeGlowing(participant);
     }
     

--- a/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/states/GameOverState.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/states/GameOverState.java
@@ -4,6 +4,7 @@ import net.kyori.adventure.audience.Audience;
 import net.kyori.adventure.text.Component;
 import org.braekpo1nt.mctmanager.Main;
 import org.braekpo1nt.mctmanager.games.game.survivalgames.SurvivalGamesGame;
+import org.braekpo1nt.mctmanager.games.game.survivalgames.SurvivalGamesParticipant;
 import org.braekpo1nt.mctmanager.games.utils.GameManagerUtils;
 import org.braekpo1nt.mctmanager.participant.Participant;
 import org.braekpo1nt.mctmanager.participant.Team;
@@ -41,7 +42,7 @@ public class GameOverState implements SurvivalGamesState {
     }
     
     @Override
-    public void onParticipantQuit(Participant participant) {
+    public void onParticipantQuit(SurvivalGamesParticipant participant) {
         context.getParticipants().remove(participant.getUniqueId());
         UUID participantUUID = participant.getUniqueId();
         String teamId = participant.getTeamId();

--- a/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/states/GameOverState.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/states/GameOverState.java
@@ -5,6 +5,7 @@ import net.kyori.adventure.text.Component;
 import org.braekpo1nt.mctmanager.Main;
 import org.braekpo1nt.mctmanager.games.game.survivalgames.SurvivalGamesGame;
 import org.braekpo1nt.mctmanager.games.game.survivalgames.SurvivalGamesParticipant;
+import org.braekpo1nt.mctmanager.games.game.survivalgames.SurvivalGamesTeam;
 import org.braekpo1nt.mctmanager.games.utils.GameManagerUtils;
 import org.braekpo1nt.mctmanager.participant.Participant;
 import org.braekpo1nt.mctmanager.participant.Team;
@@ -44,17 +45,9 @@ public class GameOverState implements SurvivalGamesState {
     @Override
     public void onParticipantQuit(SurvivalGamesParticipant participant) {
         context.getParticipants().remove(participant.getUniqueId());
-        UUID participantUUID = participant.getUniqueId();
-        String teamId = participant.getTeamId();
-        Integer oldLivingMembers = context.getLivingMembers().get(teamId);
-        if (oldLivingMembers != null) {
-            context.getLivingMembers().put(teamId, Math.max(0, oldLivingMembers - 1));
-            context.updateAliveCount(teamId);
-        }
-        context.getLivingPlayers().remove(participantUUID);
-        context.getKillCounts().remove(participantUUID);
-        context.getDeathCounts().remove(participantUUID);
-        context.getTopbar().unlinkFromTeam(participantUUID);
+        SurvivalGamesTeam team = context.getTeams().get(participant.getTeamId());
+        context.updateAliveCount(team);
+        context.getTopbar().unlinkFromTeam(participant.getUniqueId());
         resetParticipant(participant);
     }
     

--- a/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/states/GameOverState.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/states/GameOverState.java
@@ -58,11 +58,6 @@ public class GameOverState implements SurvivalGamesState {
     }
     
     @Override
-    public void initializeParticipant(Participant participant) {
-        // not used
-    }
-    
-    @Override
     public void resetParticipant(Participant participant) {
         context.resetParticipant(participant);
     }

--- a/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/states/SurvivalGamesState.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/states/SurvivalGamesState.java
@@ -1,5 +1,6 @@
 package org.braekpo1nt.mctmanager.games.game.survivalgames.states;
 
+import org.braekpo1nt.mctmanager.games.game.survivalgames.SurvivalGamesParticipant;
 import org.braekpo1nt.mctmanager.participant.Participant;
 import org.braekpo1nt.mctmanager.participant.Team;
 import org.bukkit.event.entity.EntityDamageEvent;
@@ -7,7 +8,7 @@ import org.bukkit.event.entity.PlayerDeathEvent;
 
 public interface SurvivalGamesState {
     void onParticipantJoin(Participant participant, Team team);
-    void onParticipantQuit(Participant participant);
+    void onParticipantQuit(SurvivalGamesParticipant participant);
     void resetParticipant(Participant participant);
     
     void onPlayerDamage(EntityDamageEvent event);

--- a/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/states/SurvivalGamesState.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/games/game/survivalgames/states/SurvivalGamesState.java
@@ -8,7 +8,6 @@ import org.bukkit.event.entity.PlayerDeathEvent;
 public interface SurvivalGamesState {
     void onParticipantJoin(Participant participant, Team team);
     void onParticipantQuit(Participant participant);
-    void initializeParticipant(Participant participant);
     void resetParticipant(Participant participant);
     
     void onPlayerDamage(EntityDamageEvent event);

--- a/src/main/java/org/braekpo1nt/mctmanager/participant/OfflineParticipant.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/participant/OfflineParticipant.java
@@ -2,6 +2,7 @@ package org.braekpo1nt.mctmanager.participant;
 
 import lombok.EqualsAndHashCode;
 import lombok.Getter;
+import lombok.ToString;
 import net.kyori.adventure.audience.Audience;
 import net.kyori.adventure.text.Component;
 import org.braekpo1nt.mctmanager.utils.AudienceDelegate;
@@ -11,7 +12,14 @@ import org.jetbrains.annotations.Nullable;
 
 import java.util.UUID;
 
+/**
+ * Represents a participant without a {@link Player} object. 
+ * <p>
+ * Implementations of {@link OfflineParticipant} should not use {@link EqualsAndHashCode}
+ * so that only the {@link #uniqueId} is used to check for equality. 
+ */
 @EqualsAndHashCode(onlyExplicitlyIncluded = true)
+@ToString
 public class OfflineParticipant implements AudienceDelegate {
     
     /**

--- a/src/main/java/org/braekpo1nt/mctmanager/participant/Participant.java
+++ b/src/main/java/org/braekpo1nt/mctmanager/participant/Participant.java
@@ -3,6 +3,7 @@ package org.braekpo1nt.mctmanager.participant;
 import io.papermc.paper.entity.LookAnchor;
 import lombok.EqualsAndHashCode;
 import lombok.Getter;
+import lombok.ToString;
 import net.kyori.adventure.audience.Audience;
 import net.kyori.adventure.text.Component;
 import org.braekpo1nt.mctmanager.utils.AudienceDelegate;
@@ -26,7 +27,7 @@ import java.util.stream.Collectors;
  * Represents a Participant. A participant is always a member of a {@link MCTTeam}.
  */
 @Getter
-@EqualsAndHashCode(onlyExplicitlyIncluded = true, callSuper = true)
+@ToString(callSuper = true)
 public class Participant extends OfflineParticipant implements AudienceDelegate {
     
     /**

--- a/src/test/java/org/braekpo1nt/mctmanager/participant/ParticipantTest.java
+++ b/src/test/java/org/braekpo1nt/mctmanager/participant/ParticipantTest.java
@@ -1,6 +1,9 @@
 package org.braekpo1nt.mctmanager.participant;
 
 import net.kyori.adventure.text.Component;
+import org.braekpo1nt.mctmanager.games.game.footrace.FootRaceParticipant;
+import org.braekpo1nt.mctmanager.games.game.parkourpathway.ParkourParticipant;
+import org.braekpo1nt.mctmanager.games.game.spleef.SpleefParticipant;
 import org.braekpo1nt.mctmanager.games.game.survivalgames.SurvivalGamesParticipant;
 import org.bukkit.entity.Player;
 import org.junit.jupiter.api.Assertions;
@@ -37,11 +40,35 @@ class ParticipantTest {
     }
     
     @Test
+    void parkourParticipantEquals() {
+        Participant participant = new Participant(mockPlayer, "yellow");
+        ParkourParticipant subParticipant = new ParkourParticipant(participant);
+        
+        Assertions.assertEquals(participant, subParticipant);
+    }
+    
+    @Test
+    void spleefParticipantEquals() {
+        Participant participant = new Participant(mockPlayer, "yellow");
+        SpleefParticipant subParticipant = new SpleefParticipant(participant);
+        
+        Assertions.assertEquals(participant, subParticipant);
+    }
+    
+    @Test
+    void footRaceParticipantEquals() {
+        Participant participant = new Participant(mockPlayer, "yellow");
+        FootRaceParticipant subParticipant = new FootRaceParticipant(participant, 0);
+        
+        Assertions.assertEquals(participant, subParticipant);
+    }
+    
+    @Test
     void survivalGamesParticipantEquals() {
         Participant participant = new Participant(mockPlayer, "yellow");
-        SurvivalGamesParticipant sgParticipant = new SurvivalGamesParticipant(participant);
+        SurvivalGamesParticipant subParticipant = new SurvivalGamesParticipant(participant);
         
-        Assertions.assertEquals(participant, sgParticipant);
+        Assertions.assertEquals(participant, subParticipant);
     }
     
 }

--- a/src/test/java/org/braekpo1nt/mctmanager/participant/ParticipantTest.java
+++ b/src/test/java/org/braekpo1nt/mctmanager/participant/ParticipantTest.java
@@ -1,0 +1,47 @@
+package org.braekpo1nt.mctmanager.participant;
+
+import net.kyori.adventure.text.Component;
+import org.braekpo1nt.mctmanager.games.game.survivalgames.SurvivalGamesParticipant;
+import org.bukkit.entity.Player;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+import java.util.UUID;
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+class ParticipantTest {
+    
+    static Player mockPlayer;
+    
+    @BeforeAll
+    static void setup() {
+        mockPlayer = mock(Player.class);
+        when(mockPlayer.getName()).thenReturn("rstln");
+        when(mockPlayer.getUniqueId()).thenReturn(UUID.randomUUID());
+    }
+    
+    @Test
+    void participantShouldEqualOfflineParticipant() {
+        OfflineParticipant offlineParticipant = new OfflineParticipant(
+                mockPlayer.getUniqueId(), 
+                "rstln", 
+                Component.text("rstln"), 
+                "yellow"
+        );
+        Participant participant = new Participant(offlineParticipant, mockPlayer);
+        
+        Assertions.assertEquals(offlineParticipant, participant);
+    }
+    
+    @Test
+    void survivalGamesParticipantEquals() {
+        Participant participant = new Participant(mockPlayer, "yellow");
+        SurvivalGamesParticipant sgParticipant = new SurvivalGamesParticipant(participant);
+        
+        Assertions.assertEquals(participant, sgParticipant);
+    }
+    
+}

